docs/SIL.rst

```diff
@@ -1713,9 +1713,96 @@ normal uses always before consuming uses. Any such violations of ownership
+"Ownership Kind". This lattice is defined as a 3-level lattice with::
+
+1. None being Top.
+2. Any being Bottom.
+3. All non-Any, non-None OwnershipKinds being defined as a mid-level elements of the lattice
+
+We can graphically represent the lattice via a diagram like the following::
+
+                +----+
+      +-------- |None| -----------+
+      |         +----+            |
+      |            |              |
+      v            v              v         ^
+  +-------+  +-----+------+  +---------+    |
+  | Owned |  | Guaranteed |  | Unowned |    +--- Value Ownership Kinds and
+  +-------+  +-----+------+  +---------+         Ownership Constraints
+      |            |              |
+      |            v              |         +--- Only Ownership Constraints
+      |          +---+            |         |
+      +--------->|Any|<-----------+         v
+                 +---+
+
+One moves down the lattice by performing a "meet" operation::
+
+  None meet OtherOwnershipKind -> OtherOwnershipKind
+  Unowned meet Owned -> Any
+  Owned meet Guaranteed -> Any
+
+and one moves up the lattice by performing a "join" operation, e.x.::
+
+  Any join OtherOwnershipKind -> OtherOwnershipKind
+  Owned join Any -> Owned
+  Owned join Guaranteed -> None
+
+This lattice is applied to SIL by requiring well formed SIL to:
+
+1. Define a static map of each SIL value to an OwnershipKind that classify the
+   semantics that the SIL value obeys. We call this subset of OwnershipKind to
+   be the set of `Value Ownership Kind`_: `None`_, `Unowned`_, `Guaranteed`_,
+   `Owned`_ (note conspiciously missing `Any`_). This is because in our model
+   `Any`_ represents an unknown ownership semantics and since our model is
+   statically strict, we do not allow for values to have unknown ownership.
+
+2. Define a static map from each operand of a SILInstruction, `i`, to an
+   (Ownership Kind, Boolean) called the operand's `Ownership Constraint`_. The
+   Ownership Kind element of the `Ownership Constraint`_ acts as a a static
+   "constraint" on the Value Ownership Kind that the operand's value can be mapped
+   to. The second boolean value is used to know if an operand will end the
+   lifetime of the incoming value when checking dataflow rules. The dataflow
+   rules that each `Value Ownership Kind`_ obeys is documented for each
+   `Value Ownership Kind`_ in its detailed description below.
+
+Then we take these static definitions and require that valid SIL has the
+property that given a value ``v``, an instruction ``i`` and said instruction's
+operand ``operand(i)``, that ``v`` can be used as a value in ``operand(i)`` only
+if the join of the ownership constraint of ``operand(i)`` with ``v`` is the
+ownership kind of ``v``! In symbols, we must have the following join is
+defined::
+
+  join : (OwnershipConstraint, ValueOwnershipKind) -> ValueOwnershipKind
+  OwnershipConstraint(operand(i)) join ValueOwnershipKind(v) = ValueOwnershipKind(v)
+
```

Comment:
This can be explained intuitively just by say "a value can be passed to an operand if applying the operand's ownership constraint to the value's ownership does not change the value's ownership.

Reply:
I would be fine adding that as a description. I think that this is important to visualize what it looks like algebraically.

Reply:
I added the description.